
In order to build all.img and hd_grub.img without root permissions,
we need to use mtools directly.

Signed-off-by: Thomas Backlund <tmb@mandriva.org>

diff -Nurp drakx-installer-images-1.50.orig/images/make_boot_img drakx-installer-images-1.50/images/make_boot_img
--- drakx-installer-images-1.50.orig/images/make_boot_img	2010-05-27 13:15:28.000000000 +0300
+++ drakx-installer-images-1.50/images/make_boot_img	2010-10-30 20:28:09.080983706 +0300
@@ -3,6 +3,8 @@
 use Config;
 use FileHandle;
 use MDK::Common;
+use POSIX;
+
 Config->import;
 my ($arch) = $Config{archname} =~ /(.*?)-/;
 
@@ -281,10 +283,12 @@ sub boot_img_i386 {
     _ "cp -f $rdz $tmp_mnt/$short_type.rdz";
     unlink $rdz;
 
-    my $size = max(chomp_(`du -s -k $tmp_mnt`) + 50, 1440);
+    # mtools wants the image to be a power of 32
+    my $size = max((ceil(chomp_(`du -s -k $tmp_mnt`) / 32) * 32) + 64, 1440);
     _ "dd if=/dev/zero of=$img bs=1k count=$size";
 
-    _ "mkdosfs-with-dir $tmp_mnt $img";
+    _ "/sbin/mkdosfs $img";
+    _ "mcopy -i $img $tmp_mnt/* ::";
     _ "syslinux $img";
     _ "rm -rf $tmp_mnt";
 }
@@ -775,9 +779,11 @@ sub hd_grub {
 
     my ($grub_dir) = glob("/lib/grub/*-*");
     my @grub_files = map { "$grub_dir/$_" } qw(stage1 stage2);
-    my $size = 40_000 + sum(map { -s $_ } @grub_files);
 
-    _ "dd if=/dev/zero of=$img bs=1 count=$size";
+    # mtools wants the image to be a power of 32
+    my $size = ceil((40_000 + sum(map { -s $_ } @grub_files)) / 32 / 1024) * 32;
+
+    _ "dd if=/dev/zero of=$img bs=1k count=$size";
 
     _ "rm -rf $tmp_mnt"; mkdir $tmp_mnt;
     _ "cp @grub_files $tmp_mnt";
@@ -810,7 +816,8 @@ pause .
 pause HELP END
 EOF
 
-    _ "mkdosfs-with-dir $tmp_mnt $img";
+    _ "/sbin/mkdosfs $img";
+    _ "mcopy -i $img $tmp_mnt/* ::";
     _ "rm -rf $tmp_mnt";
 
     output($mapfile, "(fd0) $img\n");
